storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -105,134 +93,99 @@ impl DBPruner for StateStorePruner {
-            self.max_index_purged_version
-                .store(min_readable_version, Ordering::Relaxed);
-            *self.index_purged_at.lock() = now;
+    pub fn prune_state_store(
+        &self,
+        min_readable_version: Version,
+        target_version: Version,
+        batch_size: usize,
+    ) -> anyhow::Result<Version> {
+        let indices_iter = StaleNodeIndicesByNumberOfItemsIterator::new(
+            &self.db,
+            min_readable_version,
+            batch_size,
+            target_version,
+        )?;
+        let indices = indices_iter.collect::<anyhow::Result<Vec<_>>>()?;
+
+        if indices.is_empty() {
+            Ok(min_readable_version)
+        } else {
+            let _timer = OTHER_TIMERS_SECONDS
+                .with_label_values(&["pruner_commit"])
+                .start_timer();
+            let new_min_readable_version =
+                indices.last().expect("Should exist.").stale_since_version;
+            let mut batch = SchemaBatch::new();
+            let mut index_batch = SchemaBatch::new();
+            // Delete stale nodes.
+            indices.into_iter().try_for_each(|index| {
+                batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key)?;
+                index_batch.delete::<StaleNodeIndexSchema>(&index)
+            })?;
+            // Delete the stale node indices.
+            self.db.write_schemas(batch)?;
+            self.db.write_schemas(index_batch)?;
+            Ok(new_min_readable_version)
         }
-        Ok(())
-    }
-
-    pub fn target_version(&self) -> Version {
-        self.target_version.load(Ordering::Relaxed)
     }
 }
-
-pub fn prune_state_store(
-    db: &DB,
-    min_readable_version: Version,
-    target_version: Version,
-    max_versions: usize,
-) -> Result<Version> {
-    let indices = StaleNodeIndicesByVersionIterator::new(db, min_readable_version, target_version)?
-        .take(max_versions) // Iterator<Item = Result<Vec<StaleNodeIndex>>>
-        .collect::<Result<Vec<_>>>()? // now Vec<Vec<StaleNodeIndex>>
-        .into_iter()
-        .flatten()
-        .collect::<Vec<_>>();
-
-    if indices.is_empty() {
-        Ok(min_readable_version)
-    } else {
-        let _timer = OTHER_TIMERS_SECONDS
-            .with_label_values(&["pruner_commit"])
-            .start_timer();
-        let new_min_readable_version = indices.last().expect("Should exist.").stale_since_version;
-        let mut batch = SchemaBatch::new();
-        indices
-            .into_iter()
-            .try_for_each(|index| batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key))?;
-        db.write_schemas(batch)?;
-        Ok(new_min_readable_version)
-    }
-}
-
-struct StaleNodeIndicesByVersionIterator<'a> {
+// This iterator will traverse `num_items` of stale nodes or all the stale nodes whose
+// `stale_since_version` = `target_min_readable_version`, whichever limit is reached first.
+struct StaleNodeIndicesByNumberOfItemsIterator<'a> {
```

Comment:
Seems to me an overkill to have such an struct? We are just iterating through the index in a pretty straightforward way? Maybe just convert to a function which returns n index items.

Reply:
Done.